File: ServiceApi.java
package com.zzflow.statesview;

import android.app.Activity;
import android.os.StrictMode;

import androidx.annotation.Nullable;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;


public class ServiceApi implements Callback {

    private Activity activity;

    public ServiceApi(Activity activity) {
        this.activity = activity;

        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder()
                .permitAll()
                .build();

        StrictMode.setThreadPolicy(policy);
    }

    /**
     * OkHttpCliente usado para as requisição
     */
    private OkHttpClient client = new OkHttpClient();

    /**
     * Quando objeto {@link Resultado}
     * for criado em {@link ServiceApi#onResultado}
     * e usado pelo onResulta o metodo
     * {@link ServiceApi#request(Request)} vai usa esse objeto
     * para passa os dados da requisição
     */
    private Resultado resultado;

    /**
     * Muito importante chama esse metodo
     * antes de fazer as requisição
     * para você obter os resultados
     *
     * @param resultado crie um new Resultado() { ... }
     */
    public void onResultado(Resultado resultado) {
        this.resultado = resultado;
    }

    /**
     * Fazer um POST
     *
     * @param url  URL do POST
     * @param body Dados para o POST
     */
    public void post(String url, String body) {
        //Obter tipo de formulario
        MediaType mediaType = checkMediaType(body);

        RequestBody requestBody = RequestBody.create(body, mediaType);

        Request request = new Request.Builder()
                .url(url)
                .post(requestBody)
                .build();

        request(request);
    }

    /**
     * Fazer um GE
     *
     * @param url URL para solicitar
     */
    public void get(String url) {
        Request request = new Request.Builder()
                .url(url)
                .build();

        request(request);
    }

    /**
     * Fazer requisição
     *
     * @param request dados para requisição
     */
    private void request(Request request) {
        client.newCall(request).enqueue(this);
    }

    /**
     * Verificar se é JSON
     *
     * @return o tipo de formulario
     */
    private MediaType checkMediaType(String body) {
        try {
            new JSONObject(body);
            // { "status": 1 }
            return MediaType.get("application/json");
        } catch (JSONException e) {
            // nome=ZzFlow&senha=1234
            return MediaType.get("application/x-www-form-urlencoded");
        }
    }

    @Override
    public void onFailure(@NotNull Call call, @NotNull IOException e) {
        prepareResponse(null);
    }

    @Override
    public void onResponse(@NotNull Call call, @NotNull final Response response) {
        prepareResponse(response);
    }

    private void prepareResponse(final Response response) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    Boolean success = response.isSuccessful();
                    String result = response.body().string();

                    resultado.onResponse(result, success);
                } catch (Exception error) {
                    resultado.onResponse(null, false);
                }
            }
        });
    }

    public interface Resultado {

        void onResponse(@Nullable String response, Boolean sucesso);
    }
}

File: ServiceApi.kt
package com.zzflow.statesview

import android.app.Activity
import android.os.StrictMode
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

class ServiceApi(private val activity: Activity) : Callback {

    private val client = OkHttpClient()
    private var resultado: Resultado? = null

    init {
        val policy = StrictMode.ThreadPolicy.Builder()
        StrictMode.setThreadPolicy(policy.permitAll().build())
    }

    fun onResultado(resultado: Resultado){
        this.resultado = resultado
    }

    fun get(url: String) {
        request(Request.Builder().url(url).build())
    }

    fun post(url: String, body: String) {
        val mediaType = checkMediaType(body)

        val requestBody: RequestBody = body.toRequestBody(mediaType)

        val build = Request.Builder()
            .url(url)
            .post(requestBody)
            .build()

        request(build)
    }

    private fun request(request: Request) {
        client.newCall(request).enqueue(this)
    }

    private fun checkMediaType(body: String): MediaType {
        return try {
            JSONObject(body)
            "application/json".toMediaType()
        } catch (e: JSONException) {
            "application/x-www-form-urlencoded".toMediaType()
        }
    }

    override fun onFailure(call: Call, e: IOException) {
        prepareResponse(null)
    }

    override fun onResponse(call: Call, response: Response) {
        prepareResponse(response)
    }

    private fun prepareResponse(response: Response?) {
        activity.runOnUiThread {
            try {
                val success = response?.isSuccessful
                val result = response?.body?.string()

                resultado?.onResponse(result, success!!)
            } catch (error: Exception) {
                resultado?.onResponse(null, false)
            }
        }
    }

    interface Resultado {

        fun onResponse(result: String?, success: Boolean)
    }
}